### PR TITLE
parquet_derive: Match fields by name, support reading selected fields rather than all

### LINKED ISSUE #6268: parquet_derive: support reading selected columns from parquet file
# Feature Description

I'm effectively using `parquet_derive` in my project, and I found that there are two inconvenient constraints:

1. The `ParquetRecordReader` enforces the struct to organize fields exactly in the **same order** in the parquet file.
2. The `ParquetRecordReader` enforces the struct to parse **all fields** in the parquet file. "all" might be exaggerating, but it is what happens if you want to get the last column, even only the last column.

As describe in its document:

> Derive flat, simple RecordReader implementations. Works by parsing a struct tagged with #[derive(ParquetRecordReader)] and emitting the correct writing code for each field of the struct. Column readers are generated in the order they are defined.

In my use cases (and I believe these are common requests), user should be able to read pruned parquet file, and they should have the freedom to re-organize fields' ordering in decoded struct.

# My Solution

I introduced a `HashMap` to map field name to its index. Of course, it assumes field name is unique, and this is always true since the current `parquet_derive` macro is applied to a flat struct without nesting.

# Pros and Cons

Obviously removing those two constraints makes `parquet_derive` a more handy tool.

But it has some implied changes:

- previously, since the `ParquetRecordReader` relies only on the index of fields, it allows that a field is named as `abc` to implicitly rename itself to `bcd` in the encoded struct. After this change, user must guarantee that the field name in `ParquetRecordReader` to exist in parquet columns.
  - I think it is more intuitive and more natural to constrain the "field name" rather than "index", if we use `ParquetRecordReader` to derive a decoder macro.
- allowing reading partial parquet file may improve the performance for some users, but introducing a `HashMap` in the parser may slowdown the function a bit.
  - when the `num_records` in a single parsing call is large enough, the cost of `HashMap` lookup is negligible.

Both implied changes seem to have a more positive impact than negative impact. Please review if this is a reasonable feature request.

### PR DESCRIPTION
**Pull request by double-free**
# Which issue does this PR close?

Closes #6268 .

# Rationale for this change

See details in the issue description.

# What changes are included in this PR?

Support reading selected fields from a parquet file.

# Are there any user-facing changes?

No API changes, but document needs to be updated:

> Derive flat, simple RecordReader implementations. Works by parsing a struct tagged with #[derive(ParquetRecordReader)] and emitting the correct writing code for each field of the struct. Column readers are generated in the order they are defined.

Column readers are generated by name instead of index now.

### CODE REVIEW

File: parquet_derive_test/src/lib.rs [hunk 1]
```diff
@@ -70,12 +70,12 @@ struct APartiallyCompleteRecord {
 // If these fields are guaranteed to be valid
 // we can load this struct into APartiallyCompleteRecord
 #[derive(PartialEq, ParquetRecordWriter, Debug)]
-struct APartiallyOptionalRecord {
+struct AnOptionalRecord {
```
Comment (alamb) [inline parquet_derive_test/src/lib.rs:73]: Why did you rename this structure? It seems like the original name `APartiallyOptionalRecord` is more accurate

File: parquet_derive/src/lib.rs [hunk 2]
```diff
@@ -206,6 +205,12 @@ pub fn parquet_record_reader(input: proc_macro::TokenStream) -> proc_macro::Toke
 
         let mut row_group_reader = row_group_reader;
 
+        // key: parquet file column name, value: column index
+        let mut name_to_index = std::collections::HashMap::new();
```
Comment (alamb) [inline parquet_derive/src/lib.rs:209]: I am not sure this strategy will work for nested structures. However, given all the tests pass we likely don't have coverage for such structures so I am not sure the existing code would work for it either. 

Thus I think this is fine for now

**Review by alamb — APPROVED** [review]
Thank you @double-free  -- the explanation and code look good to me. I am sorry for the delay in reviewing

My only question is related to renaming of the `struct` in the tests -- could you double check that and respond to my question?

Thanks again

File: parquet_derive/src/lib.rs (hunk 2 above)
Comment (double-free) [reply parquet_derive/src/lib.rs:209]: Yes, agree this is fine for now. As said in the document:

> This does not generate readers or writers for arbitrarily nested structures.

This solution is reasonable under this constraint.

File: parquet_derive_test/src/lib.rs (hunk 1 above)
Comment (double-free) [reply parquet_derive_test/src/lib.rs:73]: yes, I will restore the struct's name, though given the fact that now the columns are matched with field name (not index, as previous), so the constraint is:

> the decoded struct must have consistent field name with the column name in parquet

Given that the struct are written by `APartiallyCompleteRecord`, `maybe_i16` should be renamed back to `i16`.

**Comment by double-free** [comment]
> Thank you @double-free -- the explanation and code look good to me. I am sorry for the delay in reviewing
> 
> My only question is related to renaming of the `struct` in the tests -- could you double check that and respond to my question?
> 
> Thanks again

Hi, I have replied to those questions. Please review again.
Btw, can you point me where to update the document of `parquet_derive`? If we choose to merge the PR, this statement will be outdated:

> Column readers are generated in the order they are defined.

Column readers are generated according to names instead of indices now.

**Comment by alamb** [comment]
> Btw, can you point me where to update the document of `parquet_derive`? If we choose to merge the PR, this statement will be outdated:
> 
> > Column readers are generated in the order they are defined.

It appears from https://github.com/search?q=repo%3Aapache%2Farrow-rs%20%22order%20they%20are%22&type=code

this is controlled by https://github.com/apache/arrow-rs/blob/69e5e5f059783a4b98bd8dacb7e1487a6ae7e77b/parquet_derive/src/lib.rs#L149

Since I am working to prepare a release now, I will update the docs

**Comment by alamb** [comment]
I have updated the comments and merged up from main. I plan to merge this PR once the CI has completed

**Comment by alamb** [comment]
Thanks again @double-free

**Comment by double-free** [comment]
> Thanks again @double-free

No problem, I really appreciate your time, and will continue to contribute once I find reasonable improvements.
